protocols/messages.go

```diff
@@ -3,229 +3,80 @@ package protocols
 import (
 	"bytes"
 	"encoding/json"
-	"fmt"
 	"sort"
 
 	"github.com/statechannels/go-nitro/channel/consensus_channel"
-	"github.com/statechannels/go-nitro/channel/state"
 	"github.com/statechannels/go-nitro/payments"
 	"github.com/statechannels/go-nitro/types"
 )
 
-type PayloadType string
-
-const (
-	SignedStatePayload     PayloadType = "SignedStatePayload"
-	SignedProposalPayload  PayloadType = "SignedProposalPayload"
-	RejectionNoticePayload PayloadType = "RejectionNoticePayload"
-	VoucherPayload         PayloadType = "VoucherPayload"
-)
-
-// Message is an object to be sent across the wire. It can contain a proposal and signed states, and is addressed to a counterparty.
-type Message struct {
-	To       types.Address
-	payloads []messagePayload
-}
-
-// messagePayload is an objective id and EITHER a SignedState or SignedProposal. This package guarantees that a payload has only one value by:
-//  - validating messages that are deserialized from JSON
-//  - providing message constructors which create valid messages
-
-type messagePayload struct {
-	ObjectiveId    ObjectiveId
-	SignedState    state.SignedState
-	SignedProposal consensus_channel.SignedProposal
-	Voucher        payments.Voucher
-	Rejected       bool
-}
-
-// hasState returns true if the payload contains a signed state.
-func (p messagePayload) hasState() bool {
-	return !p.SignedState.State().Equal(state.State{})
-}
-
-// hasProposal returns true if the payload contains a signed proposal.
-func (p messagePayload) hasProposal() bool {
-	return p.SignedProposal.Proposal != consensus_channel.Proposal{}
-}
-
-// hasRejection returns true if the payload indicates that the objective is rejected
-func (p messagePayload) hasRejection() bool {
-	return p.Rejected
-}
-
-// hasVoucher returns true if the payload contains a voucher.
-func (p messagePayload) hasVoucher() bool {
-	return !(&p.Voucher).Equal(&payments.Voucher{})
-}
-
-// Type returns the type of the payload, either a SignedProposal or SignedState.
-func (p messagePayload) Type() PayloadType {
-	if p.hasProposal() {
-		return SignedProposalPayload
-	} else if p.hasState() {
-		return SignedStatePayload
-	} else if p.hasVoucher() {
-		return VoucherPayload
-	} else {
-		return RejectionNoticePayload
-	}
-}
-
-// ObjectivePayload is a struct that contains an objectiveId and EITHER a Signed State or Signed Proposal.
-type ObjectivePayload[T PayloadValue] struct {
-	Payload     T
+// ObjectivePayload is a message containing a payload of []byte that an objective is responsible for decoding.
+type ObjectivePayload struct {
+	// PayloadData is the serialized json payload
```

Comment:
I see that you are using a generic JSON encoder in the end. That seems fine for now, but still interested in hearing your general thoughts on this question ⬆️ .

Reply:
I think it would be pretty easy to allow objectives to fully decide how to encode/decode. However I think my preference would be to standardize around `json`. I can't think of a good reason a protocol would want to implement it's own encoding/decoding. 

However this is mostly based on convention right now. While the `CreateObjectivePayload` helper serializes something out to json, I think a protocol could implement their own encoding/decoding and just avoid using the CreateObjectivePayload` helper.